subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -290,10 +312,177 @@ static int tls_rx(void *ctx, unsigned char *buf, size_t len)
 	return received;
 }
 
+static int tls_add_ca_certificate(struct tls_context *tls,
+				  struct tls_credential *ca_cert)
+{
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+	int err = mbedtls_x509_crt_parse(&tls->ca_chain,
+					 ca_cert->buf, ca_cert->len);
+	if (err != 0) {
+		return -EINVAL;
+	}
+
+	return 0;
+#endif /* MBEDTLS_X509_CRT_PARSE_C */
+
+	return -ENOTSUP;
+}
+
+static void tls_set_ca_chain(struct tls_context *tls)
+{
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+	mbedtls_ssl_conf_ca_chain(&tls->config, &tls->ca_chain, NULL);
+	mbedtls_ssl_conf_authmode(&tls->config, MBEDTLS_SSL_VERIFY_REQUIRED);
+	mbedtls_ssl_conf_cert_profile(&tls->config,
+				      &mbedtls_x509_crt_profile_default);
+#endif /* MBEDTLS_X509_CRT_PARSE_C */
+}
+
+static int tls_set_own_cert(struct tls_context *tls,
+			    struct tls_credential *own_cert,
+			    struct tls_credential *priv_key)
+{
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+	int err = mbedtls_x509_crt_parse(&tls->own_cert,
+					 own_cert->buf, own_cert->len);
+	if (err != 0) {
+		return -EINVAL;
+	}
+
+	err = mbedtls_pk_parse_key(&tls->priv_key, priv_key->buf,
+				   priv_key->len, NULL, 0);
+	if (err != 0) {
+		return -EINVAL;
+	}
+
+	err = mbedtls_ssl_conf_own_cert(&tls->config, &tls->own_cert,
+					&tls->priv_key);
+	if (err != 0) {
+		err = -ENOMEM;
+	}
+
+	return 0;
+#endif /* MBEDTLS_X509_CRT_PARSE_C */
+
+	return -ENOTSUP;
+}
+
+static int tls_set_psk(struct tls_context *tls,
+		       struct tls_credential *psk,
+		       struct tls_credential *psk_id)
+{
+#if defined(MBEDTLS_KEY_EXCHANGE__SOME__PSK_ENABLED)
+	int err = mbedtls_ssl_conf_psk(&tls->config,
+				       psk->buf, psk->len,
+				       (const unsigned char *)psk_id->buf,
+				       psk_id->len - 1);
+	if (err != 0) {
+		return -EINVAL;
+	}
+
+	return 0;
+#endif
+
+	return -ENOTSUP;
+}
+
 static int tls_mbedtls_set_credentials(struct tls_context *tls)
 {
-	/* TODO Temporary solution to verify communication */
-	mbedtls_ssl_conf_authmode(&tls->config, MBEDTLS_SSL_VERIFY_NONE);
+	struct tls_credential *credential;
+	sec_tag_t tag;
+	int i, err;
+	bool tag_found, ca_cert_present = false;
+
+	credentials_lock();
+
+	for (i = 0; i < tls->options.sec_tag_list.sec_tag_count; i++) {
+		tag = tls->options.sec_tag_list.sec_tags[i];
+		credential = NULL;
+		tag_found = false;
+
+		while ((credential = credential_next_get(tag, credential))
+				!= NULL) {
+			tag_found = true;
+
+			switch (credential->type) {
+			case TLS_CREDENTIAL_CA_CERTIFICATE:
+			{
+				err = tls_add_ca_certificate(tls, credential);
+				if (err < 0) {
+					return err;
+				}
+
+				ca_cert_present = true;
+
+				break;
+			}
+
+			case TLS_CREDENTIAL_PUBLIC_CERTIFICATE:
+			{
+				struct tls_credential *priv_key =
+					credential_get(
+						tag,
+						TLS_CREDENTIAL_PRIVATE_KEY);
+
+				if (!priv_key) {
+					return -ENOENT;
```

Comment:
Don't we need to unlock the credentials before returning?

Reply:
Good point, I've completely messed up here, will fix!